### PR TITLE
Don't store column cardinality on possibly immutable column object

### PR DESCRIPTION
**Pull request by tlrobinson**
We shouldn't be storing cardinality on column objects since they might be immutable, so store them in a WeakMap instead.

### CODE REVIEW

**Comment by tlrobinson** [comment]
@metabase/core-developers Can someone review this please? Master is sort of broken for me until this is merged.

File: frontend/src/metabase/lib/api.js [hunk 1]
```diff
@@ -6,11 +6,19 @@ import EventEmitter from "events";
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/api.js]: FYI the changes to this file aren't relevant to this PR (it was to an earlier iteration), but it's something I want to add anyway.

**Comment by camsaul** [comment]
Thoughts:

1.  I can't work out what the new `Dataset.js` file has to do with this fix
2.  If this "sort of broke master" is it something we can add a test for so it doesn't get broken again?

Besides that 👍

File: frontend/src/metabase/meta/types/Dataset.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+export type ColumnName = string;
+
+// TODO: incomplete
+export type Column = {
+    name: ColumnName,
+    display_name: string,
+    base_type: string
+}
+
+export type Row = Array<string|number|Timestamp>
```
Comment (tlrobinson) [inline frontend/src/metabase/meta/types/Dataset.js]: Indeed. Anything else besides those?

Comment (tlrobinson) [inline frontend/src/metabase/meta/types/Dataset.js]: Also, it would be super cool if we could export the Clojure schemas to Flow types automagically...

**Pull request merged**